script to find differences between cov reports

rhcu opened the pull request:
Fixes #3.
 
I have finished the script to find differences in coverage information amongst two coverage reports. Before this, I collected coverage reports for 2 websites: stackoverflow.com and mozilla.org. 
I am comparing names and functions of records, and, in case, they are equal looking if the coverage information is different. There are approximately 1000 different records between reports now (out of ~18k records).

Tests are added to test/ folder with pytest framework.

marco-c commented:
We should make Travis run the pytest tests.

The diff.py script should offer a function to diff two reports, returning not only if they are different, but actually the differences. That is, given a baseline report A and a report B, it should return the things that are in B but are not in A. We don't care about things that are in A and not in B, because in the future we want to use this script to find things that are covered by the crawler (report B) but not covered in Mozilla's tests (report A).

Make the function compare two already loaded reports, so it's more generic (no opening file code and no json.load, assume the caller gives you an already loaded report).

marco-c commented:
The name of the variable for report A could be `baseline_report`. I can't come up with a good name for the second one, maybe just call it `report`.

rhcu commented:
> E.g. if the coverage for a file in A is something line [0, 1, 1, 1] and the coverage for a file in B is something like [1, 1, 1, 1], then you should return [1, 0, 0, 0].

  - Probably, [1, 1, 1, 1]? 
I've added changes so that Travis runs the tests. Functions now are cleared if not different, if different, checks whether it was executed in A or in B. If in B, then added to report. Coverage is also changed if different before adding to the report but it will result in [1, 1, 1, 1] in the above example.

marco-c commented:
> Probably, [1, 1, 1, 1]?
I've added changes so that Travis runs the tests. Functions now are cleared if not different, if different, checks whether it was executed in A or in B. If in B, then added to report. Coverage is also changed if different before adding to the report but it will result in [1, 1, 1, 1] in the above example.

No, it should be [1, 0, 0, 0].
We want the lines that are covered in B but are not covered in A.
If B is [1, 1, 1, 1] and A is [0, 1, 1, 1], then the difference between the two is [1, 0, 0, 0].
[1, 1, 1, 1] means that line 1 was covered, line 2 was covered, line 3 was covered, line 4 was covered.
[0, 1, 1, 1] means that line 1 was not covered, line 2 was covered, line 3 was covered, line 4 was covered.
So the only line which was covered in B but not covered in A was line 1.

rhcu commented:
if I have [null, null, 2, null, 5]  in A and [null, 3, 2, null, 6]  in B, I need to return [null, 3, 0, null, 6]?

marco-c commented:
> if I have [null, null, 2, null, 5] in A and [null, 3, 2, null, 6] in B, I need to return [null, 3, 0, null, 6]?

[null, 3, 0, null, 1]

The last line was executed 6 times in report B and 5 times in report A. So the difference should be that it was executed once.
In the future we should also add an option to ignore the execution count (and so the result would be '[null, 3, 0, null, 0]`), but we can do it in a follow-up.

in theory we should never have null in different positions in two reports, as `null` means it is a line which was not a coverable line (e.g. a line containing a comment). If there's a comment in a file in report A, the same will be in report B (as the reports are linked to the same source code).

rhcu commented:
Clear, thanks. Changed to be a difference between two reports: it will be 0 if they are equal and n in case if m  in A and m + n in B, respectively.

marco-c commented on .travis.yml:
```diff
@@ -6,3 +6,4 @@ install:
   - pip install -r test-requirements.txt
 script:
   - flake8 .
+  - python -m pytest test/test.py
```
You can do `pytest tests/test_*.py`, so whenever we add new test files they will be automatically executed without the need to change this.

`pytest` should be added to `test-requirements.txt` too (if it's not already there).

marco-c commented on diff.py:
```diff
@@ -0,0 +1,44 @@
+        if j == 'source_files':
+            coverage = report[j]
+    result = []
+    for i in range(0, len(baseline_coverage)):
```
Why are you using a range loop here? You can iterate `baseline_coverage` directly (same for the following line).

marco-c commented on diff.py:
```diff
@@ -0,0 +1,44 @@
+def compare_reports(baseline_report, report):
+    baseline_coverage = []
+    coverage = []
+    for k, j in zip(baseline_report, report):
```
What's the reason for this loop? Isn't it easier to just do `baseline_coverage = baseline_report['source_files']`?

marco-c commented on diff.py:
```diff
@@ -0,0 +1,44 @@
+            comp_result = compare_objects(baseline_coverage[i], coverage[j])
+            if comp_result is not None:
+                result.append(comp_result)
+    return result
```
The result should have the same format as the reports passed to compare_reports (e.g. the source files should be in `source_files`)

marco-c commented on diff.py:
```diff
@@ -0,0 +1,44 @@
+import json
+
+
+def compare_objects(obj1, obj2):
```
`compare_objects` is not a great name, too generic.

marco-c commented on diff.py:
```diff
@@ -0,0 +1,44 @@
+                    i = j - i
+                list_val.append(i)
+            obj1['coverage'] = list_val
+            return json.dumps(obj1)
```
The function should return an object, not a json string.

marco-c commented on diff.py:
```diff
@@ -0,0 +1,44 @@
+            for i, j in zip(obj1['coverage'], obj2['coverage']):
+                if j is not None and j is not 0 and j >= i:
+                    i = j - i
+                list_val.append(i)
```
If `i` is 3 and `j` is 0, the result should be `0`. With your implementation it is `3`.

marco-c commented on test/test.py:
```diff
@@ -0,0 +1,30 @@
+    assert json.loads(diff.compare_objects(obj1, obj2)) == {'branches': [], 'functions': [], 'coverage': [None, 0, None, None, 0], 'source_digest': 'another_generated_number_with_literals', 'name': 'obj-firefox/dist/include/mozilla/dom/DOMRectListBinding.h'}
+
+
+def test_compare_objects_diff_coverage_same_func():
```
We should test more cases for diff coverage.
E.g. a case where a line is covered in A but not covered in B.

marco-c commented on test/test.py:
```diff
@@ -0,0 +1,30 @@
+# -*- coding: utf-8 -*-
```
The name of the file should be `test_diff.py`, to make it clear what script it is testing.

marco-c commented on test/test.py:
```diff
@@ -0,0 +1,30 @@
+import json
+
+import diff
+
```
We should add a test for compare_reports too.

rhcu commented:
I have changed it. Should I add more tests?

marco-c commented on diff.py:
```diff
@@ -0,0 +1,44 @@
+                        if func1['name'] == func2['name']:
+                            if func1['exec'] is False and func2['exec'] is True:
+                                func1['exec'] = True
+                                diff_funcs.append(func1)
```
If you do `diff_funcs.append(func2)` you don't need the previous `func1['exec'] = True` line.

marco-c commented on diff.py:
```diff
@@ -0,0 +1,44 @@
+            list_val = []
+            for i, j in zip(obj1['coverage'], obj2['coverage']):
+                if j is not None and j is not 0 and j >= i:
+                    i = j - i
```
It's a bit confusing to reuse `i`, let's call the new variable `k`.

marco-c replied:
It might be better to specify all possible cases, to make it clearer. Something like:
```
if j is None:
    k = None
elif i is None:
    k = j
elif j > 0:
    k = j - i
```

marco-c commented:
Looks good overall, just a few minor possible improvements!

rhcu commented:
Thanks for comments. Changed. :) Test file looks quite big with formatting.

marco-c commented on diff.py:
```diff
@@ -0,0 +1,47 @@
+            list_val = []
+            for k, m in zip(obj1['coverage'], obj2['coverage']):
+                if m is None:
+                    k = None
```
When I said `i` was confusing, it was because it's better not to re-use the same name as before.
You can keep i and j for the loop, but then use another name for the diff count (if you use `i` for the first and `j` for the second, use `k` for the result).

marco-c commented on diff.py:
```diff
@@ -0,0 +1,47 @@
+                            if func1['exec'] is False and func2['exec'] is True:
+                                diff_funcs.append(func2)
+                obj1['functions'] = diff_funcs
+            list_val = []
```
No need to use `list_val`, you can use `obj2['coverage']` directly.

rhcu replied:
I thought it is not safe to modify values in the loop in Python; therefore, I've decided to make a list, where I put resulting values?

marco-c replied:
It should be safe to modify the values of the list, it isn't safe to reduce/increase the size of a list.
E.g. this isn't safe:
```
l = [3, 4, 5]
for i in l:
    l.append(i)
```

The best option here would actually be to use a list comprehension. Something like this:
```
def diff_line(i, j):
    ...

obj2['coverage'] = [diff_line(i, j) for i, j in zip(obj1['coverage'], obj2['coverage'])]
```

rhcu replied:
Ok, thanks for the explanation. Changed it now. :)
